modal/stub.py

```diff
@@ -265,15 +265,20 @@ def __getattr__(self, tag: str) -> _Object:
             # Hacky way to avoid certain issues, e.g. pickle will try to look this up
             raise AttributeError(f"Stub has no member {tag}")
         # Return a reference to an object that will be created in the future
-        return self._indexed_objects[tag]
+        obj: _Object = self._indexed_objects[tag]
+        deprecation_warning((2024, 3, 25), _Stub.__getitem__.__doc__)
```

Comment:
Do we want to gate the deprecation warning on `obj` being an `_Object`?

The main reason I ask is that I've tripped our deprecation warnings in various annoying ways, especially in tests or support code, e.g. calling `hasattr()` on an object in some duck typing scenario.

But also in Python it's not generally illegal to set / get random attributes on objects. I think it makes perfect sense to deprecate `__getitem__` / `__setitem__`, since those are not generally available on objects. And it makes sense to deprecate any side effects that `__setattr__` has that we don't want uses to rely on. But I'm not sure we need to be noisy about people accessing random attributes from a stub.

Reply:
Doesn't `__getattr__` get called only if the item isn't found in `__dict__`? So I think you can still set/get random attributes if you want to.

I'm hoping to get rid of the entire `__getattr__` method eventually, so once it's gone we'll end up with the behavior your mentioning – it will be possible to set/get arbitrary attributes. But it's going to take probably 3-6 months of deprecations to get there.

Reply:
> Doesn't __getattr__ get called only if the item isn't found in __dict__? So I think you can still set/get random attributes if you want to.

True, although the specific situation I am thinking of is a case where I was testing for the presence of an attribute (a method) the objects might not have, so it did end up taking this route.